feat: new reviewer menu actions configuration

BrayanDSO opened the pull request:
## Approach

In the commits and the learning section.

I probably should do more documentation and tests, but I'm too "in the zone" to see exactly what is missing and it will take at least a week until I got a cleaner mind to look at this with another eyes.

## How Has This Been Tested?

Android 14 phone, Samsung SM-S911B, and in the emulator with API 34 as well

https://github.com/user-attachments/assets/d7cf1bb6-cc00-4bc7-a56b-cc35273fefe3

https://github.com/user-attachments/assets/b5d5dd44-416e-4681-a035-10e8161c06bd

## Learning (optional, can help others)

The hardest part was not the drag and drop, or the scrollable menu view. It was hiding the `Bury` and `Suspend` submenus if the user only could bury a card. 

First I tried to hide the submenu arrow and override the click on the menu option. The arrow is changeable with `android:submenuArrow`, but I couldn't find a way to do that dynamically. Then I thought to recreate the menu, or take `bury` and `suspend` in account when first building it, but the result would be too messy. So I went on reading the source code of Android's menus, since the documentation is scarce.

Android menus are a bunch of interfaces with hidden implementations. 

* MenuItem -> [MenuItemImpl](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:appcompat/appcompat/src/main/java/androidx/appcompat/view/menu/MenuItemImpl.java;drc=a4a5393ead9e12a8cd9a2a6e0b27cd108dbe803d)
* Menu -> [MenuBuilder](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:appcompat/appcompat/src/main/java/androidx/appcompat/view/menu/MenuBuilder.java;l=60;drc=a4a5393ead9e12a8cd9a2a6e0b27cd108dbe803d)
* SubMenu -> [SubMenuBuilder](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:appcompat/appcompat/src/main/java/androidx/appcompat/view/menu/SubMenuBuilder.java;l=38;drc=a4a5393ead9e12a8cd9a2a6e0b27cd108dbe803d)

Overall, they are inflexible and you can't do many things with them, like adding a menu item at an specific position (you need to recreate the menu readding things in the order you want), or removing a submenu (again, you need to recreate the whole menu for that).

So if you want more power, you need to do some casting and user their implementantions, or use reflection, like I had to.

---

A toolbar menu is displayed by a [ActionMenuView](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:appcompat/appcompat/src/main/java/androidx/appcompat/widget/ActionMenuView.java;l=52;drc=a4a5393ead9e12a8cd9a2a6e0b27cd108dbe803d), which was how I managed to make a toolbar menu scrollable.

Using two ActionMenuViews is a bit hacky, but the alternatives would be doing my own implementation of a toolbar menu, either with a RecyclerView or a HorizontalScrollView (the two options for scrolling horizontally that I know of). That would be a hell of a job, specially in recreating the popups (another dark Android API that I don't want to touch again) and the tooltips. Another problem with doing that is that it wouldn't look exactly how Android toolbars look, and making things familiar to the users is important.

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [X] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/ReviewerMenuSettingsFragment.kt:
```diff
@@ -0,0 +1,21 @@
+import androidx.fragment.app.Fragment
+import com.ichi2.anki.R
+
+class ReviewerMenuSettingsFragment : Fragment(R.layout.preferences_reviewer_menu)
```
This should explain why we're using the activity, rather than just in the commit message

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/utils/Reflection.kt:
```diff
@@ -0,0 +1,38 @@
+            isAccessible = true
+        }
+    if (field == null) {
+        Timber.d("Could not find field $fieldName in class $clazz")
```
This feels like a `.w`

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/preferences/SettingsSearchBarTest.kt:
```diff
@@ -44,17 +44,17 @@ class SettingsSearchBarTest : RobolectricTest() {
         val filesResIds =
             filesToIndex.map {
-                searchItemResIdField.get(it)
+                searchItemResIdField?.get(it)
```
It was equivalent to `!!` before, could we add `requireAccessibleJavaField`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/ReviewerMenuSettingsTouchHelperCallback.kt:
```diff
@@ -0,0 +1,79 @@
+        return true
+    }
+
+    override fun onSwiped(
+        viewHolder: RecyclerView.ViewHolder,
+        direction: Int,
+    ) {
+        // do nothing
+    }
```
Nit: I'd move 'noop' methods to the end of the class, no need to break reader concentration on a nothingburger

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerFragment.kt:
```diff
@@ -126,36 +156,41 @@ class ReviewerFragment :
+            SUSPEND_NOTE -> viewModel.suspendNote()
+            SUSPEND_CARD -> viewModel.suspendCard()
+            UNSET_FLAG -> viewModel.setFlag(Flag.NONE)
+            FLAG_RED -> viewModel.setFlag(Flag.RED)
+            FLAG_BLUE -> viewModel.setFlag(Flag.BLUE)
+            FLAG_PINK -> viewModel.setFlag(Flag.PINK)
+            FLAG_TURQUOISE -> viewModel.setFlag(Flag.TURQUOISE)
+            FLAG_GREEN -> viewModel.setFlag(Flag.GREEN)
+            FLAG_ORANGE -> viewModel.setFlag(Flag.ORANGE)
+            FLAG_PURPLE -> viewModel.setFlag(Flag.PURPLE)
```
These flags are in the wrong order

david-allison reviewed (approved):
This is excellent! All comments are optional

UX: I would have dragging anywhere on the item move the item, rather than just the handle

I would have a 'tap' suggest dragging to the user

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/MenuDisplayType.kt:
```diff
@@ -0,0 +1,104 @@
+    ;
+
+    @VisibleForTesting
+    val preferenceKey get() = "ReviewerMenuDisplayType_$name"
```
Did the test

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/ReviewerOptionsFragment.kt:
```diff
@@ -31,5 +34,18 @@ class ReviewerOptionsFragment : PreferenceFragmentCompat() {
+        // An activity partially solves that, because the screen looks alright in phones, but in
+        // tablets/big screens, the preferences navigation lateral bar isn't shown.
+        requirePreference<Preference>(R.string.reviewer_menu_settings_key).setOnPreferenceClickListener {
+            val intent = SingleFragmentActivity.getIntent(requireContext(), ReviewerMenuSettingsFragment::class)
```
Just a question but did you look at embedding the preview toolbar as a preference layout through [setLayoutResource](https://developer.android.com/reference/androidx/preference/Preference#setLayoutResource(int))?

BrayanDSO replied:
The page would have two toolbars: the top one and the preview one. That doesn't look good IMO.

lukstbit reviewed (approved):
LGTM, left some minor questions.

I have to say I'm not enthusiastic about the horizontal scrollable menu container because I don't see the value. As long as menu items are hidden, I'll have to make the same effort to get to them, in one case I scroll and click menu item vs. the standard I click(overflow menu) and click menu item. The difference being that the second option is simpler to implement.

When I fixed the menu between the DeckPicker and StudyOptionsActivity I was thinking that is possible and great if we allow the user to customize the menu for each screen by calculating the available screen size(into buckets) and manipulating the menu items with _showAsAction_ (always/never) based on settings.

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/ViewerAction.kt:
```diff
@@ -0,0 +1,87 @@
+    @IdRes val menuId: Int,
+    @DrawableRes val drawableRes: Int?,
+    @StringRes val titleRes: Int = R.string.empty_string,
+    val defaultDisplayType: MenuDisplayType? = null,
```
The enum will include the controls' actions as well when gestures are implemented. Some actions like ˋPage upˋ won't have a menu entry, so they can't be ˋDISABLEDˋ. The ˋmenuIdˋ property will also be made optional by then, or converted to a broader ˋidˋ property.

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/ReviewerMenuView.kt:
```diff
@@ -0,0 +1,139 @@
+            val menuItems = MenuDisplayType.getMenuItems(context.sharedPrefs())
+            addActions(menuItems.alwaysShow, menuItems.menuOnly)
+            // wait until attached to a fragment or activity to launch the coroutine to setup flags
+            viewTreeObserver.addOnGlobalLayoutListener(
```
IIRC, I tried it but it didn't work

BrayanDSO commented:
> I'm not enthusiastic about the horizontal scrollable menu container because I don't see the value. As long as menu items are hidden, I'll have to make the same effort to get to them, in one case I scroll and click menu item vs. the standard I click(overflow menu) and click menu item.

I can see the point and I won't even going to use the feature. But, scrollable toolbars are actually quite common, as seen in many text editors and even in AnkiDroid's note editor. 

About the effort to do an action, it can be less than tapping the menu, since a person needs to do an extra tap the menu every time they use an overflown action, but if it is already scrolled to the action they want, they can keep using the action without the extra tap. 

If they want to keep using the same behavior of a default toolbar, they can. If they want to ditch the overflow menu, they can. If they want to use both, they can. All of that without making the app more complex to the users.

For us developers, it is indeed more complex than directly using a simple toolbar menu, but it isn't hard to understand or to work with at all. It's just a toolbar menu embedded in a horizontal scroll view and a separate one to show the overflow icons.

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/MenuDisplayType.kt:
```diff
@@ -0,0 +1,104 @@
+import androidx.core.content.edit
+import com.ichi2.anki.R
+
+enum class MenuDisplayType(
```
What do you mean?

david-allison replied:
Implementers choice: 

We send all preferences barring sensitive ones to ACRA, and the list is getting long. 

Do you feel knowing these values would provide help when debugging an issue? I don't feel they'd be useful and would consider adding these to the list of ignored properties

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/ReviewerMenuSettingsAdapter.kt:
```diff
@@ -0,0 +1,112 @@
+
+    private var onDragListener: ((RecyclerView.ViewHolder) -> Unit)? = null
+
+    fun setOnDragListener(onDragListener: (RecyclerView.ViewHolder) -> Unit) {
```
> it seems like you could just make the private var public

I can, but I personally prefer to not expose a property if the value isn't useful somewhere else. That also avoids exposing the listener to other classes

BrayanDSO replied:
Anyway, ended up removing it after defaulting to use long presses in the whole item

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/ReviewerMenuSettingsFragment.kt:
```diff
@@ -0,0 +1,133 @@
+
+        val recyclerViewItems =
+            listOf(
+                ReviewerMenuSettingsRecyclerItem.DisplayType(MenuDisplayType.ALWAYS),
+                *toRecyclerItems(menuItems.alwaysShow),
+                ReviewerMenuSettingsRecyclerItem.DisplayType(MenuDisplayType.MENU_ONLY),
+                *toRecyclerItems(menuItems.menuOnly),
+                ReviewerMenuSettingsRecyclerItem.DisplayType(MenuDisplayType.DISABLED),
+                *toRecyclerItems(menuItems.disabled),
```
I tried to. PTAL.

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/MenuDisplayType.kt:
```diff
@@ -0,0 +1,104 @@
+        preferences.edit { putString(preferenceKey, prefValue) }
+    }
+
+    data class Items(
```
Great idea. Thanks!

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/reviewer/MenuDisplayType.kt:
```diff
@@ -0,0 +1,104 @@
+            }
+        }
+
+        fun getMenuItems(prefs: SharedPreferences): Items {
```
> Shouldn't these two methods be in ViewerAction's companion?

I'm not sure. Both are reasonable IMO. I'm leaving as is to as a laziness move to avoid rebasing and dealing with conflicts

> Also, probably getAllMenuItems() seeing that it returns all menu items or, as an alternative, another "selection" parameter seeing that one place uses just a part of the return items:

BrayanDSO commented:
> UX: I would have dragging anywhere on the item move the item, rather than just the handle

Replaced it with a long press in the whole item. It also makes the code simpler and the list more compact, which looks better IMO

<details><summary>Screenshot</summary>

![Screenshot_20241212_222018](https://github.com/user-attachments/assets/5241988e-00e8-4526-a028-7d87345d8bd3)

</details>

---

I think I addressed all of the feedback. Thank y'all for the reviews, it helped a lot.

I'll wait to hear if anyone has more considerations to do.

david-allison reviewed (commented):
The long tap threshold is too long if the user is using the controls on the right, that should instantly start a drag action

I would consider adding a little more vertical padding between the elements, I had a few mistaps

lukstbit commented:
Still fine for me :+1:

david-allison reviewed (approved):
Animation transition from the preferences is a little wonky.

I'm more than happy though, UX of this screen is excellent!

mikehardy commented:
Fantastic @BrayanDSO
